File: algorithm/java/minimum-path-sum/P64.java
class Solution {
    public int minPathSum(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
      //  int[][] dp = new int[m][n];
        for (int i = m-1; i >= 0; i--) {
            for (int j = n-1; j >= 0; j--) {
                if (i == m-1 && j == n-1) {
                    grid[i][j] = grid[i][j];
                }  else if (i == m-1) {
                    grid[i][j] = grid[i][j] + grid[i][j+1];
                } else if (j == n-1) {
                    grid[i][j] = grid[i][j] + grid[i+1][j];
                } else {
                    grid[i][j] = grid[i][j] + Math.min(grid[i][j+1], grid[i+1][j]);
                }
            }
        }
        return grid[0][0];
    }
}

File: algorithm/go/mininum-path-sum/p64.go
package main

func minPathSum(grid [][]int) int {
	m := len(grid)
	n := len(grid[0])
	dp := make([][]int, m)
	for i := 0; i < m; i++ {
		dp[i] = make([]int, n)
	}

	for i := m - 1; i >= 0; i-- {
		for j := n - 1; j >= 0; j-- {
			if i == m-1 && j == n-1 {
				dp[i][j] = grid[i][j]
			} else if i == m-1 {
				dp[i][j] = grid[i][j] + dp[i][j+1]
			} else if j == n-1 {
				dp[i][j] = grid[i][j] + dp[i+1][j]
			} else {
				if dp[i][j+1] > dp[i+1][j] {
					dp[i][j] = grid[i][j] + dp[i+1][j]
				} else {
					dp[i][j] = grid[i][j] + dp[i][j+1]
				}
			}
		}
	}
	return dp[0][0]
}
